cpp/src/arrow/compute/kernels/codegen_internal.h

```diff
@@ -0,0 +1,648 @@
+  }
+};
+
+template <typename Type, typename Enable = void>
+struct GetValueType;
+
+template <typename Type>
+struct GetValueType<Type, enable_if_has_c_type<Type>> {
+  using T = typename Type::c_type;
+};
+
+template <typename Type>
+struct GetValueType<
+    Type, enable_if_t<is_base_binary_type<Type>::value || is_decimal_type<Type>::value ||
+                      is_fixed_size_binary_type<Type>::value>> {
+  using T = util::string_view;
+};
+
+// ----------------------------------------------------------------------
+// Generate an array kernel given template classes
+
+void ExecFail(KernelContext* ctx, const ExecBatch& batch, Datum* out);
+
+void BinaryExecFlipped(KernelContext* ctx, ArrayKernelExec exec,
+                       const ExecBatch& batch, Datum* out);
+
+// ----------------------------------------------------------------------
+// Boolean data utilities
+
+// ----------------------------------------------------------------------
+// Template kernel exec function generators
+
+template <typename T>
+void Extend(const std::vector<T>& values, std::vector<T>* out) {
+  for (const auto& t : values) {
+    out->push_back(t);
+  }
+}
+
+const std::vector<std::shared_ptr<DataType>>& BaseBinaryTypes();
+const std::vector<std::shared_ptr<DataType>>& SignedIntTypes();
+const std::vector<std::shared_ptr<DataType>>& UnsignedIntTypes();
+const std::vector<std::shared_ptr<DataType>>& IntTypes();
+const std::vector<std::shared_ptr<DataType>>& FloatingPointTypes();
+
+// Number types without boolean
+const std::vector<std::shared_ptr<DataType>>& NumericTypes();
+
+// Temporal types including time and timestamps for each unit
+const std::vector<std::shared_ptr<DataType>>& TemporalTypes();
+
+// Integer, floating point, base binary, and temporal
+const std::vector<std::shared_ptr<DataType>>& PrimitiveTypes();
+
+namespace codegen {
+
+struct SimpleExec {
+  // Operator must implement
+  //
+  // static void Call(KernelContext*, const ArrayData& in, ArrayData* out)
+  template <typename Operator>
+  static void Unary(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    if (batch[0].kind() == Datum::SCALAR) {
+      ctx->SetStatus(Status::NotImplemented("NYI"));
+    } else if (batch.length > 0) {
+      Operator::Call(ctx, *batch[0].array(), out->mutable_array());
+    }
+  }
+
+  // Operator must implement
+  //
+  // static void Call(KernelContext*, const ArrayData& arg0, const ArrayData& arg1,
+  //                  ArrayData* out)
+  template <typename Operator>
+  static void Binary(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    if (batch[0].kind() == Datum::SCALAR || batch[1].kind() == Datum::SCALAR) {
+      ctx->SetStatus(Status::NotImplemented("NYI"));
+    } else if (batch.length > 0) {
+      Operator::Call(ctx, *batch[0].array(), *batch[1].array(), out->mutable_array());
+    }
+  }
+};
+
+// TODO: Run benchmarks to determine if OutputAdapter is a zero-cost abstraction
+struct ScalarPrimitiveExec {
```

Comment:
It's difficult to understand why there's `ScalarPrimitiveExec` as well as `ScalarUnary` and other stuff.

Reply:
I will add more comments

Reply:
Added a bunch of comments to explain. If we want to create "templates to rule them all" we will need to write sufficient benchmarks to have peace of mind that all the compilers do the right thing